### ISSUE TITLE
Cannot reconnect after container restart

### ISSUE THREAD

**Issue by boramalper** [Place1/wg-access-server#66]
Hi,

First of all, thanks for this great project! I really appreciate how accessible it makes Wireguard.

- I am using Portainer + Docker to deploy wg-access-server. 
- I fail to reconnect (from my iPhone) after the container has been restarted **using an existing config**.
- If I generate a new config (i.e. add a new device) after the container has been restarted, that works perfectly well! So I cannot reconnect using a config that has been added before the restart.
- It might be worth noting that I don't have the same issue with [linuxserver/wireguard](https://hub.docker.com/r/linuxserver/wireguard/)

What might be the cause of this?

Thanks!

**iOS Client Logs**

```
2020-08-05 11:17:37.094048: [APP] startActivation: Entering (tunnel: bora iphone) 
2020-08-05 11:17:37.094809: [APP] startActivation: Tunnel is disabled. Re-enabling and saving 
2020-08-05 11:17:37.132817: [APP] startActivation: Tunnel saved after re-enabling, invoking startActivation 
2020-08-05 11:17:37.133229: [APP] startActivation: Entering (tunnel: bora iphone) 
2020-08-05 11:17:37.133628: [APP] startActivation: Starting tunnel 
2020-08-05 11:17:37.133977: [APP] startActivation: Success 
2020-08-05 11:17:37.147014: [APP] Tunnel 'bora iphone' connection status changed to 'connecting' 
2020-08-05 11:17:37.206130: [NET] App version: 0.0.20200127 (17); Go backend version: 0.0.20200121 
2020-08-05 11:17:37.206231: [NET] Starting tunnel from the app 
2020-08-05 11:17:38.046469: [NET] Tunnel interface is utun2 
2020-08-05 11:17:38.047455: [NET] DNS64: mapped xxx.xx.xxx.xxx to itself. 
2020-08-05 11:17:38.048100: [NET] Attaching to interface 
2020-08-05 11:17:38.049407: [NET] Routine: encryption worker - started 
2020-08-05 11:17:38.049547: [NET] Routine: encryption worker - started 
2020-08-05 11:17:38.049633: [NET] Routine: handshake worker - started 
2020-08-05 11:17:38.049722: [NET] Routine: decryption worker - started 
2020-08-05 11:17:38.049810: [NET] Routine: decryption worker - started 
2020-08-05 11:17:38.049894: [NET] Routine: handshake worker - started 
2020-08-05 11:17:38.049986: [NET] Routine: encryption worker - started 
2020-08-05 11:17:38.050074: [NET] Routine: decryption worker - started 
2020-08-05 11:17:38.050164: [NET] Routine: handshake worker - started 
2020-08-05 11:17:38.050262: [NET] Routine: encryption worker - started 
2020-08-05 11:17:38.050356: [NET] Routine: handshake worker - started 
2020-08-05 11:17:38.050444: [NET] Routine: handshake worker - started 
2020-08-05 11:17:38.050542: [NET] Routine: encryption worker - started 
2020-08-05 11:17:38.050628: [NET] Routine: decryption worker - started 
2020-08-05 11:17:38.050720: [NET] Routine: decryption worker - started 
2020-08-05 11:17:38.050824: [NET] Routine: encryption worker - started 
2020-08-05 11:17:38.051070: [NET] Routine: handshake worker - started 
2020-08-05 11:17:38.051160: [NET] Routine: TUN reader - started 
2020-08-05 11:17:38.051283: [NET] Routine: event worker - started 
2020-08-05 11:17:38.051347: [NET] Routine: decryption worker - started 
2020-08-05 11:17:38.051553: [NET] UAPI: Updating private key 
2020-08-05 11:17:38.052191: [NET] UAPI: Removing all peers 
2020-08-05 11:17:38.052252: [NET] UAPI: Transition to peer configuration 
2020-08-05 11:17:38.053027: [NET] peer(Sca5…Ka2o) - UAPI: Created 
2020-08-05 11:17:38.053111: [NET] peer(Sca5…Ka2o) - UAPI: Updating endpoint 
2020-08-05 11:17:38.053253: [NET] peer(Sca5…Ka2o) - UAPI: Updating persistent keepalive interval 
2020-08-05 11:17:38.053392: [NET] peer(Sca5…Ka2o) - UAPI: Removing all allowedips 
2020-08-05 11:17:38.053488: [NET] peer(Sca5…Ka2o) - UAPI: Adding allowedip 
2020-08-05 11:17:38.054832: [NET] Routine: receive incoming IPv6 - started 
2020-08-05 11:17:38.054926: [NET] Routine: receive incoming IPv4 - started 
2020-08-05 11:17:38.055028: [NET] UDP bind has been updated 
2020-08-05 11:17:38.055181: [NET] peer(Sca5…Ka2o) - Starting... 
2020-08-05 11:17:38.055504: [NET] peer(Sca5…Ka2o) - Routine: sequential receiver - started 
2020-08-05 11:17:38.055577: [NET] peer(Sca5…Ka2o) - Routine: nonce worker - started 
2020-08-05 11:17:38.055683: [NET] peer(Sca5…Ka2o) - Routine: sequential sender - started 
2020-08-05 11:17:38.055795: [NET] Device started 
2020-08-05 11:17:38.057582: [APP] Tunnel 'bora iphone' connection status changed to 'connected' 
2020-08-05 11:17:38.299137: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:17:38.303860: [NET] peer(Sca5…Ka2o) - Awaiting keypair 
2020-08-05 11:17:43.578353: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 2) 
2020-08-05 11:17:43.578739: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:17:48.746734: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 2) 
2020-08-05 11:17:48.747139: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:17:54.038063: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 2) 
2020-08-05 11:17:54.038424: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:17:59.225477: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 2) 
2020-08-05 11:17:59.225827: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:18:04.486562: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 3) 
2020-08-05 11:18:04.486949: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:18:08.979962: [APP] Status update notification timeout for tunnel 'bora iphone'. Tunnel status is now 'connected'. 
2020-08-05 11:18:09.820787: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 2) 
2020-08-05 11:18:09.820985: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:18:14.906744: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 2) 
2020-08-05 11:18:14.907174: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:18:20.218583: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 3) 
2020-08-05 11:18:20.218751: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:18:25.364612: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 2) 
2020-08-05 11:18:25.364954: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:18:30.513666: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 2) 
2020-08-05 11:18:30.514041: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:18:35.771506: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:18:41.044322: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 2) 
2020-08-05 11:18:41.044723: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:18:46.130522: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 2) 
2020-08-05 11:18:46.130913: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:18:51.245644: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 2) 
2020-08-05 11:18:51.246033: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:18:56.325986: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 2) 
2020-08-05 11:18:56.326288: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:19:01.375373: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:19:06.398608: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 2) 
2020-08-05 11:19:06.398926: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:19:11.499492: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 2) 
2020-08-05 11:19:11.499851: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:19:16.588785: [NET] peer(Sca5…Ka2o) - Handshake did not complete after 5 seconds, retrying (try 2) 
2020-08-05 11:19:16.589149: [NET] peer(Sca5…Ka2o) - Sending handshake initiation 
2020-08-05 11:19:20.896704: [APP] startDeactivation: Tunnel: bora iphone 
2020-08-05 11:19:20.903566: [APP] Tunnel 'bora iphone' connection status changed to 'disconnecting' 
2020-08-05 11:19:21.155563: [NET] Network change detected with satisfied route and interface order [en0] 
2020-08-05 11:19:21.158278: [NET] DNS64: mapped xxx.xx.xxx.xxx to itself. 
2020-08-05 11:19:21.158766: [NET] UAPI: Transition to peer configuration 
2020-08-05 11:19:21.159076: [NET] peer(Sca5…Ka2o) - UAPI: Updating endpoint 
2020-08-05 11:19:21.161552: [NET] Routine: receive incoming IPv4 - stopped 
2020-08-05 11:19:21.169471: [NET] Routine: receive incoming IPv6 - stopped 
2020-08-05 11:19:21.170954: [NET] Routine: receive incoming IPv4 - started 
2020-08-05 11:19:21.171179: [NET] Routine: receive incoming IPv6 - started 
2020-08-05 11:19:21.171788: [NET] UDP bind has been updated 
2020-08-05 11:19:21.222858: [NET] Stopping tunnel 
2020-08-05 11:19:21.223364: [NET] Device closing 
2020-08-05 11:19:21.224009: [NET] Routine: event worker - stopped 
2020-08-05 11:19:21.225276: [NET] Routine: TUN reader - stopped 
2020-08-05 11:19:21.225847: [NET] Routine: receive incoming IPv4 - stopped 
2020-08-05 11:19:21.226124: [NET] Routine: receive incoming IPv6 - stopped 
2020-08-05 11:19:21.226504: [NET] peer(Sca5…Ka2o) - Stopping... 
2020-08-05 11:19:21.226791: [NET] Routine: encryption worker - stopped 
2020-08-05 11:19:21.226986: [NET] Routine: decryption worker - stopped 
2020-08-05 11:19:21.227168: [NET] Routine: handshake worker - stopped 
2020-08-05 11:19:21.227344: [NET] Routine: encryption worker - stopped 
2020-08-05 11:19:21.227513: [NET] Routine: decryption worker - stopped 
2020-08-05 11:19:21.227677: [NET] Routine: decryption worker - stopped 
2020-08-05 11:19:21.227965: [NET] Routine: encryption worker - stopped 
2020-08-05 11:19:21.228215: [NET] Routine: handshake worker - stopped 
2020-08-05 11:19:21.228446: [NET] Routine: handshake worker - stopped 
2020-08-05 11:19:21.228631: [NET] Routine: handshake worker - stopped 
2020-08-05 11:19:21.228827: [NET] Routine: handshake worker - stopped 
2020-08-05 11:19:21.229007: [NET] Routine: decryption worker - stopped 
2020-08-05 11:19:21.229191: [NET] Routine: encryption worker - stopped 
2020-08-05 11:19:21.229372: [NET] Routine: decryption worker - stopped 
2020-08-05 11:19:21.229552: [NET] Routine: decryption worker - stopped 
2020-08-05 11:19:21.229731: [NET] Routine: handshake worker - stopped 
2020-08-05 11:19:21.229908: [NET] Routine: encryption worker - stopped 
2020-08-05 11:19:21.230081: [NET] Routine: encryption worker - stopped 
2020-08-05 11:19:21.230310: [NET] peer(Sca5…Ka2o) - Routine: nonce worker - stopped 
2020-08-05 11:19:21.230553: [NET] peer(Sca5…Ka2o) - Routine: sequential sender - stopped 
2020-08-05 11:19:21.230795: [NET] peer(Sca5…Ka2o) - Routine: sequential receiver - stopped 
2020-08-05 11:19:21.231233: [NET] Interface closed 
2020-08-05 11:19:21.240522: [APP] Tunnel 'bora iphone' connection status changed to 'disconnected'
```

**docker-compose.yaml**
```yaml
version: 2
services:
  wg-access-server:
    image: place1/wg-access-server
    container_name: wg-access-server
    cap_add:
      - NET_ADMIN
    volumes:
      - "wg-access-server-data:/data"
    ports:
      - "51820:51820/udp"
    devices:
      - "/dev/net/tun:/dev/net/tun"
    networks:
      - public_bridge  # for reverse proxy

volumes:
  wg-access-server-data:
    driver: local

networks:
  public_bridge:
    external:
      name: public_bridge
```

**Comment by Place1** [comment]
Hey there. Are you using a persistent storage backend? Check our docs for
docker, make sure you mount a volume for the data directory

**Comment by ThellraAK** [comment]
@Place1 is

"      - "wg-access-server-data:/data"    "

insufficient?

"
volumes:
  wg-access-server-data:
    driver: local
"

Is at the bottom of the docker-compose.yml as well.

**Comment by boramalper** [comment]
I also believe that using volumes shall be sufficient for persistence, and if not, it should be documented in a clearer manner. 

Any updates?

**Comment by ThellraAK** [comment]
Look at your logs from the beginning of a restart on the access server, mine showed it was regenerating the servers private key every time, causing clients to not be able to reconnect

**Comment by boramalper** [comment]
@ThellraAK Didn't you write that it caused an error?

> Using `$$(wg genkey)`
> 
> as the docker suggests (not docker-compose) results in this error
>
> ```
> level=fatal msg="failed to load wireguard config: invalid wireguard config: bad private key: wgtypes: failed to parse base64-encoded key: illegal base64 data at input byte 0" file="main.go:47"
> ```

**Comment by ThellraAK** [comment]
> @ThellraAK Didn't you write that it caused an error?
> 
> > Using `$$(wg genkey)`
> > as the docker suggests (not docker-compose) results in this error
> > ```
> > level=fatal msg="failed to load wireguard config: invalid wireguard config: bad private key: wgtypes: failed to parse base64-encoded key: illegal base64 data at input byte 0" file="main.go:47"
> > ```

Yeah you've gonna generate it yourself

Any Linux with wireguard installed on it gives you a usable output, even the container if you do the docker exec thing works for it.

**Comment by boramalper** [comment]
I see. While we can update the documentation, I believe this should be done automatically during the first run and saved to `/data` to be loaded on subsequent runs.

@Place1 Would you like to work on it yourself or me to send a PR?

**Comment by Place1** [comment]
Hey all. I've just published `v0.3.0` which includes a few changes to help address the issue outlined in this thread.

The server will now refuse to generate a private key when you're using a persistent storage backend; and as a result the docker compose template has been updated to require the private key to be configured.

I'm making this one as resolve now. Thanks everyone for assisting with the debugging process! Cheers.
